virttest/qemu_devices/qcontainer.py

```diff
@@ -1666,6 +1666,163 @@ def images_define_by_params(self, name, image_params, media=None,
+        :keyword chardev options per backend, including but not limited to:
+            port_type: port type of the serial device(optional, virtio only)
+            bus_id: the busid of parent bus(optional, virtio only)
+            nr: the nr(port) of the parent bus(optional, virtio only)
+            path: the path option of chardev device(optional)
+            name: the name option of serial device(optional)
+        :return: the device list that construct the serial device
+        """
+        if (not self.has_option('chardev') or
+                not self.has_device(serial_type)):
+            cmd = " -serial unix:'%s',server,nowait" % kwargs.get('path')
+            return qdevices.StrDev('SER-%s' % serial_id, cmdline=cmd)
+
+        devices = []
+
+        # for tcp_socket and unix_socket, both form to 'socket'
+        _backend = 'socket' if 'socket' in backend else backend
+        # Generate -chardev device
+        devices.append(qdevices.CharDevice({'backend': _backend}, serial_id))
+        chardev_id = 'chardev_%s' % serial_id
+        devices[-1].set_param('id', chardev_id)
+        devices[-1].set_param('logfile', kwargs.get('logfile'))
+        devices[-1].set_param('logappend', kwargs.get('logappend'))
+        if backend not in ['socket', 'vc', 'ringbuf',
+                           'spiceport', 'spicevmc']:
+            devices[-1].set_param('mux', kwargs.get('mux'))
+        if backend in ['file', 'pipe', 'serial',
+                       'tty', 'parallel', 'parport']:
+            devices[-1].set_param('path', kwargs.get('path'))
+        elif backend in ['spicevmc', 'spiceport']:
+            devices[-1].set_param('debug', kwargs.get('debug'))
+            devices[-1].set_param('name', kwargs.get('name'))
+        elif backend in ['stdio']:
+            devices[-1].set_param('signal', kwargs.get('signal'))
+        elif 'unix' in backend:  # unix_socket
+            devices[-1].set_param('path', kwargs.get('path'))
+        elif 'tcp' in backend or backend in ['udp']:  # tcp_socket & udp
+            devices[-1].set_param('host', kwargs.get('host'))
+            devices[-1].set_param('port', kwargs.get('port'))
+            devices[-1].set_param('ipv4', kwargs.get('ipv4'))
+            devices[-1].set_param('ipv6', kwargs.get('ipv6'))
+            if backend in ['udp']:
+                devices[-1].set_param('localaddr', kwargs.get('localaddr'))
+                devices[-1].set_param('localport', kwargs.get('localport'))
+            else:
+                devices[-1].set_param('to', kwargs.get('to'))
+                devices[-1].set_param('nodelay', kwargs.get('nodelay'))
+        if 'socket' in backend:  # tcp_socket & unix_socket
+            devices[-1].set_param('server', kwargs.get('server'))
+            devices[-1].set_param('nowait', kwargs.get('nowait'))
+            devices[-1].set_param('reconnect', kwargs.get('reconnect'))
+        devices[-1].format_params(devices[-1].params)
```

Comment:
split lines from 1691 to 1729 into function  chardev_define_by_params.

Reply:
@xutian Here I can not go through kwargs and set all params inside it for `device[-1]`, since the kwargs actually contains some other options for other devices created later, not only for this device. So that's very hard to distinguish them earlier in `define_by_params`

Reply:
@xutian I have tried to move it to `serials_define_by_params`, only including chardev related options in **kwargs, and the options for other devices is laid outside as individual parameters. Thanks for your suggestions.